File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.py
from collections import deque
def solution(n, t, m, timetable):
    answer = 100
    new_timetable = []
    for i in range(len(timetable)):
        hour, minute = timetable[i].split(":")
        time = 60 * int(hour) + int(minute)
        new_timetable.append(time)
    new_timetable.sort()
    timetable_q = deque(new_timetable)
    
    now = 9 * 60
    while n > 1:
        cnt = 0
        while cnt < m and timetable_q[0] <= now:
            cnt += 1
            timetable_q.popleft()
        n -= 1
        now += t
    if len(timetable_q) < m:
        answer = now
    else:
        tmp = 0
        for i in range(m):
            if timetable_q[i] <= now:
                tmp = max(tmp, timetable_q[i])
            else:
                tmp = now+1
                break
        answer = tmp - 1
        
    hour = answer // 60
    minute = answer % 60
    if (hour < 10):
        hour = '0' + str(hour)
    if (minute < 10):
        minute = '0' + str(minute)
    answer = str(hour) + ":" + str(minute)
    return answer

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.java
import java.util.*;
class Solution {
    public String solution(int n, int t, int m, String[] timetable) {
        String answer = "";
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (String tt : timetable) {
            int time = Integer.parseInt(tt.substring(0, 2)) * 60 + Integer.parseInt(tt.substring(3));
            pq.add(time);
        }
        int now = 9 * 60; // 현재 시간
        int last = 0; // 마지막 시간
        int total = 0; // 셔틀버스 탑승 인원
        for (int i = 0; i < n; i++) {
            total = 0;
            while (!pq.isEmpty()) {
                int cur = pq.peek();
                // 버스 출발 시간보다 일찍 오고, 버스 인원이 남아있다면
                if (cur <= now && total < m) {
                    pq.poll();
                    total++;
                } else {
                    break;
                }
                last = cur;
            }
            now += t;
            
        }
        // 마지막 셔틀에 탄 인원이 m명보다 작으면 해당 셔틀 도착시간에 탈 수 있음.
        if (total < m) {
            last = now - t;
        } else {
            last--; // 실제로 가장 마지막에 탄 인원의 시간을 저장해뒀기 때문에 그 사람보다 1분 빨라야 탈 수 있음.
        }
        
        String hour = String.format("%02d", last / 60);
        String minute = String.format("%02d", last % 60);
        return hour + ":" + minute;
    }
}